Blog post: React debounce and throttle with hooks (#8)
divyanshu013/blog: content/blog/react-debounce-throttle-hooks/index.md

---
title: 'React debounce and throttle with hooks'
description: 'How to use debounce and throttle and abstract them into hooks'
date: 2020-07-03
categories: ['react']
ogImage: ./og-image.png
---

[Hooks](https://reactjs.org/docs/hooks-intro.html) are a brilliant addition to React. They simplify a lot of logic that had to be earlier split up into different lifecycles with `class` components. They do, however, require a _different_ mental model, [especially for timers](https://overreacted.io/making-setinterval-declarative-with-react-hooks/).

## Debounce and throttle

There are a ton of blog posts around debounce and throttle so I won't be diving into how to write our own debounce and throttle. For brevity, consider [`debounce`](https://lodash.com/docs/4.17.15#debounce) and [`throttle`](https://lodash.com/docs/4.17.15#throttle) from Lodash.

> If you need a quick refresher, both accept a (callback) function and a _delay_ in milliseconds (say `x`) and returns another function with some special behavior:

Reviewer: ```suggestion
If you need a quick refresher, both accept a (callback) function and a _delay_ in milliseconds (say `x`) and return another function with some special behavior:
```

Author: Nice catch man, I need a grammar refresher 😄

## useCallback

Although [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback) is commonly used for performance optimizations when passing callbacks to child components, we can use its constraint of memoizing a callback function to ensure the `debouncedSave` references the same debounced function across renders.

This works as expected:

```js
import React, { useState, useCallback } from 'react';
import debounce from 'lodash.debounce';

function App() {
	const [value, setValue] = useState('');
	const [dbValue, saveToDb] = useState(''); // would be an API call normally

	// highlight-starts
	const debouncedSave = useCallback(
		debounce(nextValue => saveToDb(nextValue), 1000),
		[], // will be created only once initially
	);
	// highlight-ends

	const handleChange = event => {
		const { value: nextValue } = event.target;
		setValue(nextValue);
		// Even though handleChange is created on each render and executed
		// it references the same debouncedSave that was created initially
		debouncedSave(nextValue);
	};

	return <main>{/* Same as before */}</main>;
}
```

## useRef

> [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) gives us a mutable object whose `current` property refers to the passed initial value. If we don't change it manually, the value would persist for the entire lifetime of the component. This is similar to class instance fields (as we had earlier with classes, defining methods and properties on `this`).

Reviewer: ```suggestion
[`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) gives us a mutable object whose `current` property refers to the passed initial value. If we don't change it manually, the value would persist for the entire lifetime of the component. This is similar to class instance properties (i.e. defining methods and properties on `this`).
```

This also works as expected:

```js
import React, { useState, useRef } from 'react';
import debounce from 'lodash.debounce';

function App() {
	const [value, setValue] = useState('');
	const [dbValue, saveToDb] = useState(''); // would be an API call normally

	// This remains same across renders
	// highlight-starts
	const debouncedSave = useRef(debounce(nextValue => saveToDb(nextValue), 1000))
		.current;
	// highlight-ends

	const handleChange = event => {
		const { value: nextValue } = event.target;
		setValue(nextValue);
		// Even though handleChange is created on each render and executed
		// it references the same debouncedSave that was created initially
		debouncedSave(nextValue);
	};

	return <main>{/* Same as before */}</main>;
}
```

## Extracting a custom hook

> Both the above examples using `useCallback` and `useRef` works fine for our usecase. This is fine for one off cases but wouldn't it be nice if there was a simpler API? It would be nice if we don't have to think about putting these callbacks in `useCallback` and `useRef`. We can certainly do so by abstracting this logic into a custom `useDebounce` hook!

Reviewer: ```suggestion
Both of the above examples using `useCallback` and `useRef` work fine for our usecase. This is good for one off cases but wouldn't it be nice if there was a simpler API? The code will be much more readable if we don't have to assemble the debounce logic with `useCallback` and `useRef`. We can certainly do so by abstracting this logic into a custom `useDebounce` hook!
```

	// highlight-starts
	const debouncedSave = useDebounce(nextValue => saveToDb(nextValue), 1000);
	// highlight-ends

	const handleChange = event => {
		const { value: nextValue } = event.target;
		setValue(nextValue);
		debouncedSave(nextValue);
	};

	return <main>{/* Same as before */}</main>;
}
```

This code also works as expected but weirdly my TypeScript linter throws an error:

```
React Hook useCallback received a function whose dependencies are unknown.
Pass an inline function instead. eslint(react-hooks/exhaustive-deps)
```

The same works fine without any linting errors in JavaScript (using a `create-react-app` template). Anyway, here's an alternative for the same `useDebounce` hook written using `useRef`:

```js
function useDebounce(callback, delay) {
	// Memoizing the callback because if it's an arrow function
	// it would be different on each render
	const memoizedCallback = useCallback(callback, []);
	const debouncedFn = useRef(debounce(memoizedCallback, delay));

	useEffect(() => {
		debouncedFn.current = debounce(memoizedCallback, delay);
	}, [memoizedCallback, debouncedFn, delay]);

	return debouncedFn.current;
}
```

> This isn't as clean as the previous example with `useCallback` but I think perhaps the error I'm getting is a bug and would be fixed in the TypeScript linter soon.

Reviewer: ```suggestion
This isn't as clean as the previous example with `useCallback`, but perhaps the lint warning I'm getting is a bug and will probably be fixed in the TypeScript linter soon.
```
